src/main/core/worker.rs

```diff
@@ -135,77 +148,80 @@ impl Worker {
+        debug_assert!(old.is_none());
     }
 
     /// Clear the currently-active Thread.
     pub fn clear_active_thread() {
-        WORKER.with(|worker| {
-            let mut worker = worker.get().unwrap().borrow_mut();
-            let old = worker.active_thread_info.take();
-            debug_assert!(!old.is_none());
-        });
+        let old = Worker::with_mut(|w| w.active_thread_info.take());
+        debug_assert!(!old.is_none());
     }
 
     /// Whether currently running on a live Worker.
     pub fn is_alive() -> bool {
-        unsafe { cshadow::worker_isAlive() != 0 }
-    }
-
-    /// Current simulation time, or None if not running on a live Worker.
-    pub fn current_time() -> Option<SimulationTime> {
-        if !Worker::is_alive() {
-            return None;
-        }
-        SimulationTime::from_c_simtime(unsafe { cshadow::worker_getCurrentTime() })
+        Worker::with(|_| ()).is_some()
     }
 
     /// ID of this thread's Worker, if any.
     pub fn thread_id() -> Option<WorkerThreadID> {
-        WORKER.with(|worker| worker.get().map(|w| w.borrow().worker_id))
+        Worker::with(|w| w.worker_id)
     }
 
     pub fn active_process_native_pid() -> Option<nix::unistd::Pid> {
-        WORKER
-            .with(|worker| {
-                worker.get().map(|w| {
-                    w.borrow()
-                        .active_process_info
-                        .as_ref()
-                        .map(|p| p.native_pid)
-                })
-            })
-            .flatten()
+        Worker::with(|w| w.active_process_info.as_ref().map(|p| p.native_pid)).flatten()
     }
 
     pub fn active_process_id() -> Option<ProcessId> {
-        WORKER
-            .with(|worker| {
-                worker
-                    .get()
-                    .map(|w| w.borrow().active_process_info.as_ref().map(|p| p.id))
-            })
-            .flatten()
+        Worker::with(|w| w.active_process_info.as_ref().map(|p| p.id)).flatten()
     }
 
     pub fn active_thread_native_tid() -> Option<nix::unistd::Pid> {
+        Worker::with(|w| w.active_thread_info.as_ref().map(|t| t.native_tid)).flatten()
+    }
+
+    fn set_round_end_time(t: SimulationTime) {
+        Worker::with_mut(|w| w.clock.barrier.replace(t)).unwrap();
+    }
+
+    fn round_end_time() -> Option<SimulationTime> {
+        Worker::with(|w| w.clock.barrier).flatten()
+    }
+
+    fn set_current_time(t: SimulationTime) {
+        Worker::with_mut(|w| w.clock.now.replace(t)).unwrap();
+    }
+
+    fn clear_current_time() {
+        Worker::with_mut(|w| w.clock.now.take());
+    }
+
+    pub fn current_time() -> Option<SimulationTime> {
+        Worker::with(|w| w.clock.now).flatten()
+    }
+
+    fn set_last_event_time(t: SimulationTime) {
+        Worker::with_mut(|w| w.clock.last.replace(t)).unwrap();
+    }
+
+    fn with<F, O>(f: F) -> Option<O>
+    where
+        F: FnOnce(&Worker) -> O,
```

Comment:
Might be useful to have a comment explaining when this function returns `Some` and when it returns `None`.

Reply:
Done